### PR TITLE
Mcrypt disabled warning in Two Factor authentication

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/9507.
### Summary of Changes

This solves the php fatal error in https://github.com/joomla/joomla-cms/issues/9507 by adding a warning message when mcrypt php extension is not enabled.
### Testing Instructions

Before applying this PR patch is the issue described in https://github.com/joomla/joomla-cms/issues/9507.

Test if 2FA still works as before.
1. Install Joomla with latest staging
2. Apply patch
3. Disable mcrypt and restart server
4. Test configuring 2FA (see below) - you will get a warning message.
5. Enable mcrypt and restart server
6. Test configuring 2FA (see below)
7. Now logout and try to login with 2FA
8. Code review

Note: Test Frontend (profile) and Backend user edit.
### How to test 2FA
- Enable Google Authenticator two factor authentication plugin: Extensions > Plugins (twofactorauth)
- Create a dummy user and configure it to use with Google Authenticator: Users > Manage > Edit (2FA tab)

Note: you can use a chrome plugin for configuring and using 2FA, for instance https://chrome.google.com/webstore/detail/authenticator/bhghoamapcdpbohphigoooaddinpkbai
#### Observations

Other PR related:
- https://github.com/joomla/joomla-cms/pull/9532
- https://github.com/joomla/joomla-cms/pull/9508

### CODE REVIEW

**Comment by brianteeman** [comment]
1. Shouldnt it be an error not a warning
2. How about "The mcrypt php extension is needed before you can enable Two Factor Authentication."
3. Please alpha-order the new string

**Comment by andrepereiradasilva** [comment]
done, done and done

**Comment by andrepereiradasilva** [comment]
I noticed now that we still need to add this check in the frontend too.

**Comment by andrepereiradasilva** [comment]
ok done

File: administrator/components/com_users/views/user/tmpl/edit.php [hunk 1]
```diff
@@ -144,7 +145,15 @@
 			<div class="clearfix"></div>
 			<?php endif; ?>
 		</fieldset>
-
+		<?php else : ?>
+		<div id="system-message-container">
```
Comment (mbabker) [inline administrator/components/com_users/views/user/tmpl/edit.php]: This probably shouldn't be duplicating the `system-message-container` div...

Comment (andrepereiradasilva) [inline administrator/components/com_users/views/user/tmpl/edit.php]: thanks. right again.
It could make a scenario with two same element ids in HTML...

**Comment by mbabker** [comment]
Looks fine to me...

**Comment by zero-24** [comment]
I have just send you a CS fix. This is on my testing list for this afternoon ;) https://github.com/andrepereiradasilva/joomla-cms/pull/43<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by zero-24** [comment]
sorry one more andrepereiradasilva#44

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 7a6bbfa9a57b5d1ebf3ba85a82157a278fe79bf0<br /><br />Good one. Thanks. To the tester please note that the message is now an error and not a warning as on the screenshot.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by Schmidie64** [comment]
@icampus 
I have tested this item unsuccessfully.
I wasn't able to reproduce the warning. 
After i install Google Authenticator Two Factor Authentication everything worked well and i not get a warning popup.

**Comment by brianteeman** [comment]
@Schmidie64 are you sure you were able to disable mcrypt? Did phpinfo confirm it? I know I was never able to successfully disable it on my system

**Comment by Schmidie64** [comment]
I have tested this item :red_circle: unsuccessfully on 7a6bbfa9a57b5d1ebf3ba85a82157a278fe79bf0<br /><br />@icampus 
 I have tested this item unsuccessfully.
 I wasn't able to reproduce the warning. 
 After i install Google Authenticator Two Factor Authentication everything worked well and i not get a warning popup.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by mxkmp29** [comment]
I tested @icampus

How did you disable mcrypt on your system ?
I tried it on linux and had no success. 

With mcrypt everything is working fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by andrepereiradasilva** [comment]
go to your php.ini loook for this
```ini
extension=mcrypt.so
```
comment it
```ini
; extension=mcrypt.so
```

restart your apache/php or php-fpm server

**Comment by mxkmp29** [comment]
The php.ini doesn't contain this line.

"_[mcrypt]
; For more information about mcrypt settings see http://php.net/mcrypt-module-open

; Directory where to load mcrypt algorithms
; Default: Compiled in into libmcrypt (usually /usr/local/lib/libmcrypt)
;mcrypt.algorithms_dir=

; Directory where to load mcrypt modes
; Default: Compiled in into libmcrypt (usually /usr/local/lib/libmcrypt)
;mcrypt.modes_dir=_"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by andrepereiradasilva** [comment]
search for `mcrypt` in your php ini files, you will find something like `extension = mcrypt.something`  it can deffer across Operating systems. That is the line you have to comment

**Comment by mxkmp29** [comment]
I found it on php.ini-pre1.7.2 and it was already comment

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by brianteeman** [comment]
Managed to disable mcrypt this time not sure what I was doing wrong before and confirmed the fatal error

Unfortunately after applying the patch I still get the following fatal errors when I try to edit the user

````[02-Aug-2016 13:31:54 Europe/London] PHP Notice:  Use of undefined constant MCRYPT_RIJNDAEL_256 - assumed 'MCRYPT_RIJNDAEL_256' in /Applications/MAMP/htdocs/github-joomla-cms/libraries/fof/encrypt/aes.php on line 44
[02-Aug-2016 13:31:54 Europe/London] PHP Notice:  Use of undefined constant MCRYPT_MODE_CBC - assumed 'MCRYPT_MODE_CBC' in /Applications/MAMP/htdocs/github-joomla-cms/libraries/fof/encrypt/aes.php on line 63
[02-Aug-2016 13:31:54 Europe/London] PHP Fatal error:  Call to undefined function mcrypt_get_iv_size() in /Applications/MAMP/htdocs/github-joomla-cms/libraries/fof/encrypt/aes.php on line 144````<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by andrepereiradasilva** [comment]
maybe it's because of the conflicts... i need to fix and check first
will try to do this tonight.

**Comment by brianteeman** [comment]
@andrepereiradasilva could you check this please - also as fof has been updated you might not even need this PR anymore<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by mbabker** [comment]
Still needed as it's adding checks for a scenario when `mcrypt` isn't installed or enabled on the system.  Just needs rebased first to see where things are at.

**Comment by andrepereiradasilva** [comment]
yes i can solve the conflict, but know the twofacto auth exists in some more places. 
have to check then all later.

**Comment by andrepereiradasilva** [comment]
conflicts fixed.
ready to test.

**Comment by andrepereiradasilva** [comment]
ok, sorry for that, but needed to solve additional issues.

Ready for tests

@brianteeman can you test?

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 979004b64fbbea0c7383d6c18f02b1e3ba82cf2c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9548">issues.joomla.org/joomla-cms/9548</a>.</sub>

**Comment by andrepereiradasilva** [comment]
thanks @brianteeman

File: administrator/language/en-GB/en-GB.com_users.ini [hunk 2]
```diff
@@ -365,4 +365,5 @@ COM_USERS_VIEW_NEW_USER_TITLE="Users: New"
 COM_USERS_VIEW_NOTES_TITLE="User Notes"
 COM_USERS_VIEW_USERS_TITLE="Users"
 COM_USERS_XML_DESCRIPTION="Component for managing users"
+COM_USERS_WARNING_MCRYPT_NEEDED="The mcrypt php extension is needed before you can enable Two Factor Authentication."
```
Comment (mbabker) [inline administrator/language/en-GB/en-GB.com_users.ini]: "mcrypt PHP extension", or "mcrypt extension for PHP".  Either way PHP needs to be capitalized.

File: language/en-GB/en-GB.com_users.ini [hunk 3]
```diff
@@ -162,3 +162,4 @@ COM_USERS_USER_FIELD_TIMEZONE_DESC="Choose your time zone."
 COM_USERS_USER_FIELD_TIMEZONE_LABEL="Time Zone"
 COM_USERS_USER_NOT_FOUND="User not found."
 COM_USERS_USER_SAVE_FAILED="Failed to save user: %s"
+COM_USERS_WARNING_MCRYPT_NEEDED="The mcrypt php extension is needed before you can enable Two Factor Authentication."
```
Comment (mbabker) [inline language/en-GB/en-GB.com_users.ini]: Same here.

**Comment by andrepereiradasilva** [comment]
coorected text string as @mbabker comments

**Comment by wilsonge** [comment]
https://github.com/joomla/joomla-cms/pull/11673 This should have been fixed with this FOF update

**Comment by andrepereiradasilva** [comment]
yes, it seems fixed. so no needs for mcrypt anymore right?

**Comment by andrepereiradasilva** [comment]
@wilsonge @mbabker if joomla now does not need mcrypt so https://github.com/joomla/joomla-cms/pull/9508 AND https://github.com/joomla/joomla-cms/pull/9532 should be reverted, and of course, this closed.

**Comment by wilsonge** [comment]
I guess it depends if the other libraries michael mentioned have since been patched?

**Comment by andrepereiradasilva** [comment]
ok, i see there is an issue for that https://github.com/joomla/joomla-cms/issues/11398

**Pull request closed without merging**